sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ApplyFunctionExpression.scala

```diff
@@ -30,10 +30,11 @@ case class ApplyFunctionExpression(
   override def dataType: DataType = function.resultType()
 
   private lazy val reusedRow = new GenericInternalRow(children.size)
+  private lazy val childrenWithIndex = children.zipWithIndex
 
   /** Returns the result of evaluating this expression on a given input Row */
   override def eval(input: InternalRow): Any = {
-    children.zipWithIndex.foreach {
+    childrenWithIndex.foreach {
```

Comment:
shall we simply use a while loop with a counter? Adding a lazy val as a class member variable looks a bit heavy.

Reply:
I'm not sure if this is very useful. After the change the cost of the loop is negligible:

<img width="806" alt="Screen Shot 2021-05-11 at 10 56 15 AM" src="https://user-images.githubusercontent.com/506679/117926975-6c8a7900-b2ae-11eb-9f2b-9b40f0873115.png">

Plus I think `foreach` is easier to read.

Reply:
Actually @sunchao it's in the style guide: https://github.com/databricks/scala-style-guide#traversal-and-zipwithindex. Looks like here's a hot path.

Reply:
I think it's worthwhile trying.

Reply:
Sure I can give it a try later if you have strong opinion on this.

Reply:
thanks man!

Reply:
Running benchmarks now. Let's see.